doc/manpages/ocsptool.1
.TH ocsptool 1 "15 Apr 2014" "@VERSION@" "User Commands"
.\"
.\"  DO NOT EDIT THIS FILE   (ocsptool-args.man)
.\"
.\"  It has been AutoGen-ed  April 15, 2014 at 12:47:20 PM by AutoGen 5.18.1
.\"  From the definitions    ../../src/ocsptool-args.def.tmp
.\"  and the template file   agman-cmd.tpl
.\"
.SH NAME
ocsptool \- GnuTLS OCSP tool
.SH SYNOPSIS
.B ocsptool
.\" Mixture of short (flag) options and long options
.RB [ \-\fIflag\fP " [\fIvalue\fP]]... [" \-\-\fIopt\-name\fP "[[=| ]\fIvalue\fP]]..."
.PP
All arguments must be options.
.PP
.SH "DESCRIPTION"
Ocsptool is a program that can parse and print information about
OCSP requests/responses, generate requests and verify responses.
.sp
.SH "OPTIONS"
.TP
.BR  \-d " \fInumber\fP, " \-\-debug "=" \fInumber\fP
Enable debugging.
This option takes an integer number as its argument.
The value of \fInumber\fP is constrained to being:
.in +4
.nf
.na
in the range  0 through 9999
.fi
.in -4
.sp
Specifies the debug level.
.TP
.BR  \-V ", " \-\-verbose
More verbose output.
This option may appear an unlimited number of times.
.sp
.TP
.BR  \-\-infile "=\fIfile\fP"
Input file.
.sp
.TP
.BR  \-\-outfile "=\fIstring\fP"
Output file.
.sp
.TP
.BR  \-\-ask "[=\fIserver name|url\fP]"
Ask an OCSP/HTTP server on a certificate validity.
This option must appear in combination with the following options:
load-cert, load-issuer.
.sp
Connects to the specified HTTP OCSP server and queries on the validity of the loaded certificate.
.TP
.BR  \-e ", " \-\-verify\-response
Verify response.
.sp
.TP
.BR  \-i ", " \-\-request\-info
Print information on a OCSP request.
.sp
.TP
.BR  \-j ", " \-\-response\-info
Print information on a OCSP response.
.sp
.TP
.BR  \-q ", " \-\-generate\-request
Generate an OCSP request.
.sp
.TP
.BR  \-\-nonce, " \fB\-\-no\-nonce\fP"
Use (or not) a nonce to OCSP request.
The \fIno\-nonce\fP form will disable the option.
.sp
.TP
.BR  \-\-load\-issuer "=\fIfile\fP"
Read issuer certificate from file.
.sp
.TP
.BR  \-\-load\-cert "=\fIfile\fP"
Read certificate to check from file.
.sp
.TP
.BR  \-\-load\-trust "=\fIfile\fP"
Read OCSP trust anchors from file.
This option must not appear in combination with any of the following options:
load-signer.
.sp
.TP
.BR  \-\-load\-signer "=\fIfile\fP"
Read OCSP response signer from file.
This option must not appear in combination with any of the following options:
load-trust.
.sp
.TP
.BR  \-\-inder, " \fB\-\-no\-inder\fP"
Use DER format for input certificates and private keys.
The \fIno\-inder\fP form will disable the option.
.sp
.TP
.BR  \-Q " \fIfile\fP, " \-\-load\-request "=" \fIfile\fP
Read DER encoded OCSP request from file.
.sp
.TP
.BR  \-S " \fIfile\fP, " \-\-load\-response "=" \fIfile\fP
Read DER encoded OCSP response from file.
.sp
.TP
.BR \-h , " \-\-help"
Display usage information and exit.
.TP
.BR \-! , " \-\-more-help"
Pass the extended usage information through a pager.
.TP
.BR \-v " [{\fIv|c|n\fP}]," " \-\-version" "[={\fIv|c|n\fP}]"
Output version of program and exit.  The default mode is `v', a simple
version.  The `c' mode will print copyright information and `n' will
print the full copyright notice.
.SH EXAMPLES
.br
\fBPrint information about an OCSP request\fP
.br
.sp
To parse an OCSP request and print information about the content, the
\fB\-i\fP or \fB\-\-request\-info\fP parameter may be used as follows.
The \fB\-Q\fP parameter specify the name of the file containing the
OCSP request, and it should contain the OCSP request in binary DER
format.
.sp
.br
.in +4
.nf
$ ocsptool \-i \-Q ocsp\-request.der
.in -4
.fi
.sp
The input file may also be sent to standard input like this:
.sp
.br
.in +4
.nf
$ cat ocsp\-request.der | ocsptool \-\-request\-info
.in -4
.fi
.sp
.br
\fBPrint information about an OCSP response\fP
.br
.sp
Similar to parsing OCSP requests, OCSP responses can be parsed using
the \fB\-j\fP or \fB\-\-response\-info\fP as follows.
.sp
.br
.in +4
.nf
$ ocsptool \-j \-Q ocsp\-response.der
$ cat ocsp\-response.der | ocsptool \-\-response\-info
.in -4
.fi
.sp
.br
\fBGenerate an OCSP request\fP
.br
.sp
The \fB\-q\fP or \fB\-\-generate\-request\fP parameters are used to
generate an OCSP request.  By default the OCSP request is written to
standard output in binary DER format, but can be stored in a file
using \fB\-\-outfile\fP.  To generate an OCSP request the issuer of the
certificate to check needs to be specified with \fB\-\-load\-issuer\fP
and the certificate to check with \fB\-\-load\-cert\fP.  By default PEM
format is used for these files, although \fB\-\-inder\fP can be used to
specify that the input files are in DER format.
.sp
.br
.in +4
.nf
$ ocsptool \-q \-\-load\-issuer issuer.pem \-\-load\-cert client.pem \
           \-\-outfile ocsp\-request.der
.in -4
.fi
.sp
When generating OCSP requests, the tool will add an OCSP extension
containing a nonce.  This behaviour can be disabled by specifying
\fB\-\-no\-nonce\fP.
.sp
.br
\fBVerify signature in OCSP response\fP
.br
.sp
To verify the signature in an OCSP response the \fB\-e\fP or
\fB\-\-verify\-response\fP parameter is used.  The tool will read an
OCSP response in DER format from standard input, or from the file
specified by \fB\-\-load\-response\fP.  The OCSP response is verified
against a set of trust anchors, which are specified using
\fB\-\-load\-trust\fP.  The trust anchors are concatenated certificates
in PEM format.  The certificate that signed the OCSP response needs to
be in the set of trust anchors, or the issuer of the signer
certificate needs to be in the set of trust anchors and the OCSP
Extended Key Usage bit has to be asserted in the signer certificate.
.sp
.br
.in +4
.nf
$ ocsptool \-e \-\-load\-trust issuer.pem \
           \-\-load\-response ocsp\-response.der
.in -4
.fi
.sp
The tool will print status of verification.
.sp
.br
\fBVerify signature in OCSP response against given certificate\fP
.br
.sp
It is possible to override the normal trust logic if you know that a
certain certificate is supposed to have signed the OCSP response, and
you want to use it to check the signature.  This is achieved using
\fB\-\-load\-signer\fP instead of \fB\-\-load\-trust\fP.  This will load
one certificate and it will be used to verify the signature in the
OCSP response.  It will not check the Extended Key Usage bit.
.sp
.br
.in +4
.nf
$ ocsptool \-e \-\-load\-signer ocsp\-signer.pem \
           \-\-load\-response ocsp\-response.der
.in -4
.fi
.sp
This approach is normally only relevant in two situations.  The first
is when the OCSP response does not contain a copy of the signer
certificate, so the \fB\-\-load\-trust\fP code would fail.  The second
is if you want to avoid the indirect mode where the OCSP response
signer certificate is signed by a trust anchor.
.sp
.br
\fBReal\-world example\fP
.br
.sp
Here is an example of how to generate an OCSP request for a
certificate and to verify the response.  For illustration we'll use
the \fBblog.josefsson.org\fP host, which (as of writing) uses a
certificate from CACert.  First we'll use \fBgnutls\-cli\fP to get a
copy of the server certificate chain.  The server is not required to
send this information, but this particular one is configured to do so.
.sp
.br
.in +4
.nf
$ echo | gnutls\-cli \-p 443 blog.josefsson.org \-\-print\-cert > chain.pem
.in -4
.fi
.sp
Use a text editor on \fBchain.pem\fP to create three files for each
separate certificates, called \fBcert.pem\fP for the first
certificate for the domain itself, secondly \fBissuer.pem\fP for the
intermediate certificate and \fBroot.pem\fP for the final root
certificate.
.sp
The domain certificate normally contains a pointer to where the OCSP
responder is located, in the Authority Information Access Information
extension.  For example, from \fBcerttool \-i < cert.pem\fP there is
this information:
.sp
.br
.in +4
.nf
Authority Information Access Information (not critical):
Access Method: 1.3.6.1.5.5.7.48.1 (id\-ad\-ocsp)
Access Location URI: http://ocsp.CAcert.org/
.in -4
.fi
.sp
This means the CA support OCSP queries over HTTP.  We are now ready to
create a OCSP request for the certificate.
.sp
.br
.in +4
.nf
$ ocsptool \-\-ask ocsp.CAcert.org \-\-load\-issuer issuer.pem \
           \-\-load\-cert cert.pem \-\-outfile ocsp\-response.der
.in -4
.fi
.sp
The request is sent via HTTP to the OCSP server address specified. If the
address is ommited ocsptool will use the address stored in the certificate.
.sp
.SH "EXIT STATUS"
One of the following exit values will be returned:
.TP
.BR 0 " (EXIT_SUCCESS)"
Successful program execution.
.TP
.BR 1 " (EXIT_FAILURE)"
The operation failed or the command syntax was not valid.
.TP
.BR 70 " (EX_SOFTWARE)"
libopts had an internal operational error.  Please report
it to autogen-users@lists.sourceforge.net.  Thank you.
.SH "SEE ALSO"
    certtool (1)
.SH "AUTHORS"
Nikos Mavrogiannopoulos, Simon Josefsson and others; see /usr/share/doc/gnutls/AUTHORS for a complete list.
.SH "COPYRIGHT"
Copyright (C) 2000-@YEAR@ Free Software Foundation, and others all rights reserved.
This program is released under the terms of the GNU General Public License, version 3 or later.
.SH "BUGS"
Please send bug reports to: @PACKAGE_BUGREPORT@
.SH "NOTES"
This manual page was \fIAutoGen\fP-erated from the \fBocsptool\fP
option definitions.


NEWS
GnuTLS NEWS -- History of user-visible changes.                -*- outline -*-
Copyright (C) 2000-2012 Free Software Foundation, Inc.
See the end for copying conditions.

** libgnutls: Added OCSP support.
There is a new header file gnutls/ocsp.h and a set of new functions
under the gnutls_ocsp namespace.  Currently the functionality provided
is to parse and extract information from OCSP requests/responses, to
generate OCSP requests and to verify OCSP responses.  See the manual
for more information.  Run ./configure with --disable-ocsp to build
GnuTLS without OCSP support.

** ocsptool: Added new command line tool.
The tool can parse OCSP request/responses, generate OCSP requests and
verify OCSP responses.  See the manual for more information.

** API and ABI modifications:
gnutls/ocsp.h: Added new header file.
gnutls_ocsp_print_formats_t: Added new type.
gnutls_ocsp_resp_status_t: Added new type.
gnutls_ocsp_cert_status_t: Added new type.
gnutls_x509_crl_reason_t: Added new type.
gnutls_ocsp_req_add_cert: Added.
gnutls_ocsp_req_add_cert_id: Added.
gnutls_ocsp_req_deinit: Added.
gnutls_ocsp_req_export: Added.
gnutls_ocsp_req_get_cert_id: Added.
gnutls_ocsp_req_get_extension: Added.
gnutls_ocsp_req_get_nonce: Added.
gnutls_ocsp_req_get_version: Added.
gnutls_ocsp_req_import: Added.
gnutls_ocsp_req_init: Added.
gnutls_ocsp_req_print: Added.
gnutls_ocsp_req_randomize_nonce: Added.
gnutls_ocsp_req_set_extension: Added.
gnutls_ocsp_req_set_nonce: Added.
gnutls_ocsp_resp_deinit: Added.
gnutls_ocsp_resp_export: Added.
gnutls_ocsp_resp_get_certs: Added.
gnutls_ocsp_resp_get_extension: Added.
gnutls_ocsp_resp_get_nonce: Added.
gnutls_ocsp_resp_get_produced: Added.
gnutls_ocsp_resp_get_responder: Added.
gnutls_ocsp_resp_get_response: Added.
gnutls_ocsp_resp_get_signature: Added.
gnutls_ocsp_resp_get_signature_algorithm: Added.
gnutls_ocsp_resp_get_single: Added.
gnutls_ocsp_resp_get_status: Added.
gnutls_ocsp_resp_get_version: Added.
gnutls_ocsp_resp_import: Added.
gnutls_ocsp_resp_init: Added.
gnutls_ocsp_resp_print: Added.
gnutls_ocsp_resp_verify: Added.

* Version 3.0.12 (unreleased)

** certtool: --outder option now works for private
and public keys as well.

** libgnutls: Added error code GNUTLS_E_NO_PRIORITIES_WERE_SET
to warn when no or insufficient priorities were set.

** libgnutls: Corrected an alignment issue in ECDH
key generation which prevented some keys from being
correctly aligned in rare circumstances.

** libgnutls: Corrected memory leaks in DH parameter
generation and ecc_projective_check_point().

** libgnutls: Added gnutls_x509_dn_oid_name() to 
return a descriptive name of a DN OID.

** API and ABI modifications:
gnutls_pubkey_encrypt_data: Added
gnutls_x509_dn_oid_name: Added
gnutls_session_resumption_requested: Added

* Version 3.0.11 (released 2012-01-06)

** libgnutls: Corrected functionality of 
gnutls_record_get_direction(). Reported by Philip Allison.

** libgnutls: Provide less timing information when decoding
TLS/DTLS record packets. Patch by Nadhem Alfardan.

** API and ABI modifications:
No changes since last version.


* Version 3.0.10 (released 2012-01-04)

** gnutls-cli/serv: Set don't fragment bit in DTLS sessions
in Linux as well as in BSD.

** gnutls-cli: Fixed reading from windows terminals.

** libgnutls: When GNUTLS_OPENPGP_FMT_BASE64 is specified
the stream is assumed to be base64 encoded (previously
the encoding was auto-detected). This avoids a decoding
issue in windows systems.

** libgnutls: Corrected ciphersuite GNUTLS_ECDHE_PSK_AES_256_CBC_SHA384

** libgnutls: Added ciphersuites: GNUTLS_PSK_WITH_AES_256_GCM_SHA384
and GNUTLS_DHE_PSK_WITH_AES_256_GCM_SHA384.

** libgnutls: Added function gnutls_random_art() to convert 
fingerprints to images (currently ascii-art).

** libgnutls: Corrected bug in DSA private key parsing, which
prevented the verification of the key.

** API and ABI modifications:
gnutls_random_art: Added


* Version 3.0.9 (released 2011-12-13)

** certtool: Added new parameter --dh-info.

** certtool: -l option was overloaded so if combined with --priority
it will only list the ciphersuites that are enabled by the given
priority string.

** libgnutls: Added new priority string %SERVER_PRECEDENCE, which
changes the ciphersuite selection procedure. If specified the server
priorities will be used for selection instead of the client's.

** libgnutls: Optimizations in Diffie-Hellman parameters generation
and key exchange.

** libgnutls: When session tickets are negotiated and used in a
session, a server will not store that session data into its cache.

** libgnutls: Added the SECP192R1 curve.

** libgnutls: Added gnutls_priority_get_cipher_suite_index() to
allow listing the ciphersuites enabled in a priority structure.
It outputs an index to be used in gnutls_get_cipher_suite_info().

** libgnutls: Optimizations in the elliptic curve code --timing
attacks resistant code is only used in ECDSA private key operations.

** doc: man pages for API functions generation was fixed and are
now added again in the distribution.

** API and ABI modifications:
GNUTLS_ECC_CURVE_SECP192R1: New curve definition
gnutls_priority_get_cipher_suite_index: Added


* Version 3.0.8 (released 2011-11-12)

** certtool: Certtool -e returns error code on verification
failure.

** certtool: Verifies parameters of generated keys.

** libgnutls: Corrected ECC key generation (introduced in 3.0.6)

** libgnutls: Provide less timing information when decoding
TLS/DTLS record packets.

** doc: man pages for API functions were removed.
The reason was that the code that auto-generated the man pages missed
many APIs and we couldn't fix it (volunteers welcome).  See the info
manual or the GTK-DOC manual instead.

** API and ABI modifications:
gnutls_x509_privkey_verify_params: Added


* Version 3.0.7 (released 2011-11-08)

** libgnutls: Corrected fix in gnutls_session_get_data()
to report the actual session size when the provided buffer
is not enough.


src/ocsptool.c
/*
 * Copyright (C) 2011 Free Software Foundation, Inc.
 *
 * This file is part of GnuTLS.
 *
 * GnuTLS is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * GnuTLS is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see
 * <http://www.gnu.org/licenses/>.
 */

#include <config.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#include <gnutls/gnutls.h>
#include <gnutls/ocsp.h>
#include <gnutls/x509.h>
#include <gnutls/crypto.h>

/* Gnulib portability files. */
#include <error.h>
#include <progname.h>
#include <version-etc.h>
#include <read-file.h>

#include <ocsptool-common.h>
#include <ocsptool-gaa.h>

gaainfo info;
FILE *outfile;
FILE *infile;

static void
tls_log_func (int level, const char *str)
{
  fprintf (stderr, "|<%d>| %s", level, str);
}

static void
request_info (void)
{
  gnutls_ocsp_req_t req;
  int ret;
  gnutls_datum_t dat;
  size_t size;

  ret = gnutls_ocsp_req_init (&req);
  if (ret < 0)
    error (EXIT_FAILURE, 0, "ocsp_req_init: %s", gnutls_strerror (ret));

  if (info.req)
    dat.data = read_binary_file (info.req, &size);
  else
    dat.data = fread_file (infile, &size);
  if (dat.data == NULL)
    error (EXIT_FAILURE, errno, "reading request");
  dat.size = size;

  ret = gnutls_ocsp_req_import (req, &dat);
  free (dat.data);
  if (ret < 0)
    error (EXIT_FAILURE, 0, "importing request: %s", gnutls_strerror (ret));

  ret = gnutls_ocsp_req_print (req, GNUTLS_OCSP_PRINT_FULL, &dat);
  if (ret != 0)
    error (EXIT_FAILURE, 0, "ocsp_req_print: %s", gnutls_strerror (ret));

  printf ("%.*s", dat.size, dat.data);
  gnutls_free (dat.data);

  gnutls_ocsp_req_deinit (req);
}

static void
response_info (void)
{
  gnutls_ocsp_resp_t resp;
  int ret;
  gnutls_datum_t dat;
  size_t size;

  ret = gnutls_ocsp_resp_init (&resp);
  if (ret < 0)
    error (EXIT_FAILURE, 0, "ocsp_resp_init: %s", gnutls_strerror (ret));

  if (info.resp)
    dat.data = read_binary_file (info.resp, &size);
  else
    dat.data = fread_file (infile, &size);
  if (dat.data == NULL)
    error (EXIT_FAILURE, errno, "reading response");
  dat.size = size;

  ret = gnutls_ocsp_resp_import (resp, &dat);
  free (dat.data);
  if (ret < 0)
    error (EXIT_FAILURE, 0, "importing response: %s", gnutls_strerror (ret));

  ret = gnutls_ocsp_resp_print (resp, GNUTLS_OCSP_PRINT_FULL, &dat);
  if (ret != 0)
    error (EXIT_FAILURE, 0, "ocsp_resp_print: %s", gnutls_strerror (ret));

  printf ("%.*s", dat.size, dat.data);
  gnutls_free (dat.data);

  gnutls_ocsp_resp_deinit (resp);
}

static gnutls_x509_crt_t
load_issuer (void)
{
  gnutls_x509_crt_t crt;
  int ret;
  gnutls_datum_t dat;
  size_t size;

  if (info.issuer == NULL)
    error (EXIT_FAILURE, 0, "missing --load-issuer");

  ret = gnutls_x509_crt_init (&crt);
  if (ret < 0)
    error (EXIT_FAILURE, 0, "crt_init: %s", gnutls_strerror (ret));

  dat.data = read_binary_file (info.issuer, &size);
  dat.size = size;

  if (!dat.data)
    error (EXIT_FAILURE, errno, "reading --load-issuer: %s", info.issuer);

  ret = gnutls_x509_crt_import (crt, &dat, info.inder);
  free (dat.data);
  if (ret < 0)
    error (EXIT_FAILURE, 0, "importing --load-issuer: %s: %s",
           info.issuer, gnutls_strerror (ret));

  return crt;
}

static gnutls_x509_crt_t
load_cert (void)
{
  gnutls_x509_crt_t crt;
  int ret;
  gnutls_datum_t dat;
  size_t size;

  if (info.cert == NULL)
    error (EXIT_FAILURE, 0, "missing --load-cert");

  ret = gnutls_x509_crt_init (&crt);
  if (ret < 0)
    error (EXIT_FAILURE, 0, "crt_init: %s", gnutls_strerror (ret));

  dat.data = read_binary_file (info.cert, &size);
  dat.size = size;

  if (!dat.data)
    error (EXIT_FAILURE, errno, "reading --load-cert: %s", info.cert);

  ret = gnutls_x509_crt_import (crt, &dat, info.inder);
  free (dat.data);
  if (ret < 0)
    error (EXIT_FAILURE, 0, "importing --load-cert: %s: %s",
           info.cert, gnutls_strerror (ret));

  return crt;
}

static void
generate_request (void)
{
  gnutls_ocsp_req_t req;
  int ret;
  gnutls_datum_t dat;
  gnutls_x509_crt_t issuer, cert;

  ret = gnutls_ocsp_req_init (&req);
  if (ret < 0)
    error (EXIT_FAILURE, 0, "ocsp_req_init: %s", gnutls_strerror (ret));

  issuer = load_issuer ();
  cert = load_cert ();

  ret = gnutls_ocsp_req_add_cert (req, GNUTLS_DIG_SHA1,
				  issuer, cert);
  if (ret < 0)
    error (EXIT_FAILURE, 0, "ocsp_req_add_cert: %s", gnutls_strerror (ret));

  gnutls_x509_crt_deinit (cert);
  gnutls_x509_crt_deinit (issuer);

  if (!info.nononce)
    {
      char noncebuf[23];
      gnutls_datum_t nonce = { noncebuf, sizeof (noncebuf) };

      ret = gnutls_rnd (GNUTLS_RND_RANDOM, nonce.data, nonce.size);
      if (ret < 0)
	error (EXIT_FAILURE, 0, "gnutls_rnd: %s", gnutls_strerror (ret));

      ret = gnutls_ocsp_req_set_nonce (req, 0, &nonce);
      if (ret < 0)
	error (EXIT_FAILURE, 0, "ocsp_req_set_nonce: %s",
	       gnutls_strerror (ret));
    }

  ret = gnutls_ocsp_req_export (req, &dat);
  if (ret != 0)
    error (EXIT_FAILURE, 0, "ocsp_req_export: %s", gnutls_strerror (ret));

  fwrite (dat.data, 1, dat.size, outfile);

  gnutls_free (dat.data);

  gnutls_ocsp_req_deinit (req);
}

static void
print_verify_res (unsigned int output)
{
  int comma = 0;

  if (output)
    {
      printf ("Failure");
      comma = 1;
    }
  else
    {
      printf ("Success");
      comma = 1;
    }

  if (output & GNUTLS_OCSP_VERIFY_SIGNER_NOT_FOUND)
    {
      if (comma)
        printf (", ");
      printf ("Signer cert not found");
      comma = 1;
    }

  if (output & GNUTLS_OCSP_VERIFY_SIGNER_KEYUSAGE_ERROR)
    {
      if (comma)
        printf (", ");
      printf ("Signer cert keyusage error");
      comma = 1;
    }

  if (output & GNUTLS_OCSP_VERIFY_UNTRUSTED_SIGNER)
    {
      if (comma)
        printf (", ");
      printf ("Signer cert is not trusted");
      comma = 1;
    }

  if (output & GNUTLS_OCSP_VERIFY_INSECURE_ALGORITHM)
    {
      if (comma)
        printf (", ");
      printf ("Insecure algorithm");
      comma = 1;
    }

  if (output & GNUTLS_OCSP_VERIFY_SIGNATURE_FAILURE)
    {
      if (comma)
        printf (", ");
      printf ("Signature failure");
      comma = 1;
    }

  if (output & GNUTLS_OCSP_VERIFY_CERT_NOT_ACTIVATED)
    {
      if (comma)
        printf (", ");
      printf ("Signer cert not yet activated");
      comma = 1;
    }

  if (output & GNUTLS_OCSP_VERIFY_CERT_EXPIRED)
    {
      if (comma)
        printf (", ");
      printf ("Signer cert expired");
      comma = 1;
    }
}

static void
verify_response (void)
{
  gnutls_ocsp_resp_t resp;
  int ret;
  gnutls_datum_t dat;
  size_t size;
  gnutls_x509_crt_t *x509_ca_list = NULL;
  unsigned int x509_ncas = 0;
  gnutls_x509_trust_list_t list;
  gnutls_x509_crt_t signer;
  unsigned verify;

  ret = gnutls_ocsp_resp_init (&resp);
  if (ret < 0)
    error (EXIT_FAILURE, 0, "ocsp_resp_init: %s", gnutls_strerror (ret));

  if (info.resp)
    dat.data = read_binary_file (info.resp, &size);
  else
    dat.data = fread_file (infile, &size);
  if (dat.data == NULL)
    error (EXIT_FAILURE, errno, "reading response");
  dat.size = size;

  ret = gnutls_ocsp_resp_import (resp, &dat);
  free (dat.data);
  if (ret < 0)
    error (EXIT_FAILURE, 0, "importing response: %s", gnutls_strerror (ret));

  if (info.trust && info.signer)
    error (EXIT_FAILURE, 0, "cannot mix --load-trust and --load-signer");
  else if (info.signer == NULL)
    {
      dat.data = read_binary_file (info.trust, &size);
      if (dat.data == NULL)
	error (EXIT_FAILURE, errno, "reading --load-trust: %s", info.trust);
      dat.size = size;

      ret = gnutls_x509_trust_list_init (&list, 0);
      if (ret < 0)
	error (EXIT_FAILURE, 0, "gnutls_x509_trust_list_init: %s",
	       gnutls_strerror (ret));

      ret = gnutls_x509_crt_list_import2 (&x509_ca_list, &x509_ncas, &dat,
					  GNUTLS_X509_FMT_PEM, 0);
      if (ret < 0 || x509_ncas < 1)
	error (EXIT_FAILURE, 0, "error parsing CAs: %s",
	       gnutls_strerror (ret));

      if (info.verbose)
	{
	  unsigned int i;
	  for (i = 0; i < x509_ncas; i++)
	    {
	      gnutls_datum_t out;

	      ret = gnutls_x509_crt_print (x509_ca_list[i],
					   GNUTLS_CRT_PRINT_ONELINE, &out);
	      if (ret < 0)
		error (EXIT_FAILURE, 0, "gnutls_x509_crt_print: %s",
		       gnutls_strerror (ret));

	      printf ("Trust anchor %d: %.*s\n", i, out.size, out.data);
	      gnutls_free (out.data);
	    }
	}

      ret = gnutls_x509_trust_list_add_cas (list, x509_ca_list, x509_ncas, 0);
      if (ret < 0)
	error (EXIT_FAILURE, 0, "gnutls_x509_trust_add_cas: %s",
	       gnutls_strerror (ret));

      if (info.verbose)
	fprintf (stdout, "Loaded %d trust anchors\n", x509_ncas);

      ret = gnutls_ocsp_resp_verify (resp, list, &verify, 0);
      if (ret < 0)
	error (EXIT_FAILURE, 0, "gnutls_ocsp_resp_verify: %s",
	       gnutls_strerror (ret));
    }
  else if (info.trust == NULL)
    {
      ret = gnutls_x509_crt_init (&signer);
      if (ret < 0)
	error (EXIT_FAILURE, 0, "crt_init: %s", gnutls_strerror (ret));

      dat.data = read_binary_file (info.signer, &size);
      if (dat.data == NULL)
	error (EXIT_FAILURE, errno, "reading --load-signer: %s", info.signer);
      dat.size = size;

      ret = gnutls_x509_crt_import (signer, &dat, info.inder);
      free (dat.data);
      if (ret < 0)
	error (EXIT_FAILURE, 0, "importing --load-signer: %s: %s",
	       info.signer, gnutls_strerror (ret));

      if (info.verbose)
	{
	  gnutls_datum_t out;

	  ret = gnutls_x509_crt_print (signer, GNUTLS_CRT_PRINT_ONELINE, &out);
	  if (ret < 0)
	    error (EXIT_FAILURE, 0, "gnutls_x509_crt_print: %s",
		   gnutls_strerror (ret));

	  printf ("Signer: %.*s\n", out.size, out.data);
	  gnutls_free (out.data);
	}

      ret = gnutls_ocsp_resp_verify_direct (resp, signer, &verify, 0);
      if (ret < 0)
	error (EXIT_FAILURE, 0, "gnutls_ocsp_resp_verify_direct: %s",
	       gnutls_strerror (ret));
    }
  else
    error (EXIT_FAILURE, 0, "missing --load-trust or --load-signer");

  printf ("Verifying OCSP Response: ");
  print_verify_res (verify);
  printf (".\n");

  gnutls_ocsp_resp_deinit (resp);
}

int
main (int argc, char **argv)
{
  int ret;

  set_program_name (argv[0]);

  if ((ret = gnutls_global_init ()) < 0)
    error (EXIT_FAILURE, 0, "global_init: %s", gnutls_strerror (ret));

  if (gaa (argc, argv, &info) != -1)
    {
      fprintf (stderr, "Try `%s --help' for more information.\n",
               program_name);
      exit (EXIT_FAILURE);
    }

  gnutls_global_set_log_function (tls_log_func);
  gnutls_global_set_log_level (info.debug);

  if (info.outfile)
    {
      outfile = fopen (info.outfile, "wb");
      if (outfile == NULL)
        error (EXIT_FAILURE, errno, "%s", info.outfile);
    }
  else
    outfile = stdout;

  if (info.infile)
    {
      infile = fopen (info.infile, "rb");
      if (infile == NULL)
        error (EXIT_FAILURE, errno, "%s", info.infile);
    }
  else
    infile = stdin;

  if (info.inder)
    info.inder = GNUTLS_X509_FMT_DER;
  else
    info.inder = GNUTLS_X509_FMT_PEM;

  switch (info.action)
    {
    case ACTION_REQ_INFO:
      request_info ();
      break;

    case ACTION_RESP_INFO:
      response_info ();
      break;

    case ACTION_GEN_REQ:
      generate_request ();
      break;

    case ACTION_VERIFY_RESP:
      verify_response ();
      break;

    default:
      gaa_help();
    }

  return 0;
}

void
ocsptool_version (void)
{
  const char *p = PACKAGE_NAME;
  if (strcmp (gnutls_check_version (NULL), PACKAGE_VERSION) != 0)
    p = PACKAGE_STRING;
  version_etc (stdout, "ocsptool", p, gnutls_check_version (NULL),
               "Simon Josefsson", (char *) NULL);
}
